File: frontend/src/lib/api/pin.ts
import * as PinFacade from "../../../bindings/ben/desktop/pinfacade";
import {
  Types,
  type PinState,
  type PinSubjectRef,
  type JobSnapshot,
} from "./models";
import { traceWailsCall } from "@/lib/observability/trace";

export const PIN_PROFILE = "desktop";

export function startPin(subject: PinSubjectRef, profile = PIN_PROFILE) {
  return traceWailsCall("pin", "start_pin", { profile, subject }, () =>
    PinFacade.StartPin(
      new Types.PinIntentRequest({
        Profile: profile,
        Subject: subject,
      }),
    ),
  ) as Promise<JobSnapshot>;
}

export function unpin(subject: PinSubjectRef, profile = PIN_PROFILE) {
  return traceWailsCall("pin", "unpin", { profile, subject }, () =>
    PinFacade.Unpin(
      new Types.PinIntentRequest({
        Profile: profile,
        Subject: subject,
      }),
    ),
  );
}

export function getPinState(subject: PinSubjectRef, profile = PIN_PROFILE) {
  return traceWailsCall("pin", "get_pin_state", { profile, subject }, () =>
    PinFacade.GetPinState(
      new Types.PinStateRequest({
        Profile: profile,
        Subject: subject,
      }),
    ),
  ) as Promise<PinState>;
}

export function listPinStates(
  subjects: PinSubjectRef[],
  profile = PIN_PROFILE,
): Promise<PinState[]> {
  return traceWailsCall(
    "pin",
    "list_pin_states",
    { profile, count: subjects.length },
    () =>
      PinFacade.ListPinStates(
        new Types.PinStateListRequest({
          Profile: profile,
          Subjects: subjects,
        }),
      ),
  );
}

export function subscribePinEvents() {
  return PinFacade.SubscribePinEvents();
}

File: api/types/pin.go
package apitypes

import (
	"context"
	"time"
)

type PinSubjectKind string

const (
	PinSubjectRecordingCluster PinSubjectKind = "recording_cluster"
	PinSubjectRecordingVariant PinSubjectKind = "recording_variant"
	PinSubjectAlbumVariant     PinSubjectKind = "album_variant"
	PinSubjectPlaylist         PinSubjectKind = "playlist"
	PinSubjectLikedPlaylist    PinSubjectKind = "liked_playlist"
)

type PinSubjectRef struct {
	Kind PinSubjectKind
	ID   string
}

type PinIntentRequest struct {
	Subject PinSubjectRef
	Profile string
}

type PinStateRequest struct {
	Subject PinSubjectRef
	Profile string
}

type PinStateListRequest struct {
	Subjects []PinSubjectRef
	Profile  string
}

type PinSourceRef struct {
	Subject PinSubjectRef
	Direct  bool
	Profile string
}

type PinState struct {
	Subject PinSubjectRef
	Pinned  bool
	Direct  bool
	Covered bool
	Pending bool
	Sources []PinSourceRef
}

type PinChangeEvent struct {
	InvalidateAll bool
	Subjects      []PinSubjectRef
	OccurredAt    time.Time
}

// PinSurface exposes the durable pin state model. Job-starting APIs live on
// the desktop runtime layer because they return desktop job snapshots.
type PinSurface interface {
	Unpin(ctx context.Context, req PinIntentRequest) error
	ListPinStates(ctx context.Context, req PinStateListRequest) ([]PinState, error)
	GetPinState(ctx context.Context, req PinStateRequest) (PinState, error)
}
